### PR TITLE
Cli with encryption

### LINKED ISSUE #953: Update CLI to support wallet encryption

### LINKED ISSUE #1171: Update CLI to support encrypted wallets
Wallet CLI password handling:

Password arguments can be specified with a cli flag.  
If a password is provided and the wallet is not encrypted, this is an error.
If a password is not provided and the wallet is encrypted, prompt the user for a password.

Prompting for a password:

https://godoc.org/golang.org/x/crypto/ssh/terminal#ReadPassword
    - Make sure this does not need cgo to build (add CGO_ENABLED=0 to makefile test commands?)
    - If it does, maybe try https://github.com/bgentry/speakeasy 

Wallet CLI changes:

`addPrivateKey`
    - Add password argument

`createRawTransaction`
    - Add password argument

`generateAddresses`
    - Add password argument

`generateWallet`
    - Add encrypt (bool) and password arguments

`send`
    - Add password argument

New command:

`showSeed`

Args:
    - wallet file from env vars
    - password, if encrypted

Print the seed for the wallet on the command line.  As opposed to `GET /wallet/seed`, this doesn't care if the wallet is encrypted or not.

New command:

`decryptWallet`

Args:
    - wallet file from env vars
    - password

Return an error if the wallet is not encrypted (exit code 1), otherwise if successful return nothing (exit code 0)

New command:

`encryptWallet`

Args:
    - wallet file from env vars
    - password

Return an error if the wallet is already encrypted.

### PR DESCRIPTION
**Pull request by iketheadore**
Fixes #1171 

Changes:
- Add password argument to `addPrivateKey`, `createRawTransaction`, `generateAddresses`, `generateWallet`, `send`.
- Add `showSeed` command with encryption support
- Add `decryptWallet` command
- Add `decryptWallet` command
- Update partial integration tests

TODO:
- [x] Add CLI integration tests for `showSeed`
- [x] Add CLI integration tests for `decryptWallet`
- [x] Fill CLI integration tests for `encryptWallet`
- [x] Update CHANGELOG.md

Does this change need to mentioned in CHANGELOG.md?
Yes

### CODE REVIEW

File: src/api/cli/cli.go [hunk 1]
```diff
@@ -305,3 +309,72 @@ func printJSON(obj interface{}) error {
 
 	return nil
 }
+
+// readPasswordFromTerminal promotes user to enter password and read it.
+func readPasswordFromTerminal() ([]byte, error) {
+	// Promotes to enter the wallet password
+	fmt.Fprint(os.Stdout, "enter password:")
+	bp, err := terminal.ReadPassword(0)
```
Comment (gz-c) [inline src/api/cli/cli.go]: should we use os.Stdin?

File: src/api/cli/cli.go [hunk 2]
```diff
@@ -305,3 +309,72 @@ func printJSON(obj interface{}) error {
...
+func (w WalletLoadError) Error() string {
+	return w.error.Error()
+}
+
+// WalletSaveError is returned if a wallet could not be saved
+type WalletSaveError struct {
+	error
+}
+
+func (w WalletSaveError) Error() string {
```
Comment (gz-c) [inline src/api/cli/cli.go]: isn't this redundant?

File: src/api/cli/generate_wallet.go [hunk 3]
```diff
@@ -120,11 +131,49 @@ func generateWallet(c *gcli.Context) error {
...
-	wlt, err := GenerateWallet(wltName, label, sd, num)
+
+	cryptoType, err := wallet.CryptoTypeFromString(c.String("x"))
+	if err != nil {
+		return err
+	}
+
+	pr := NewPasswordReader([]byte(c.String("p")))
+	switch pr.(type) {
+	case nil:
```
Comment (gz-c) [inline src/api/cli/generate_wallet.go]: pr won't be nil?

File: src/api/cli/add_private_key.go [hunk 4]
```diff
@@ -90,13 +91,42 @@ func AddPrivateKey(wlt *wallet.Wallet, key string) error {
...
+		return WalletLoadError{err}
+	}
+
+	switch pr.(type) {
+	case nil:
+		if wlt.IsEncrypted() {
+			return wallet.ErrMissingPassword
+		}
+	case PasswordFromBytes:
+		p, _ := pr.Password()
```
[the remark this answers is not in the document]
Comment (iketheadore) [reply src/api/cli/add_private_key.go]: `PasswordFromBytes` type doesn't have error.

File: src/api/cli/cli.go (hunk 1 above)
Comment (iketheadore) [reply src/api/cli/cli.go]: `os.Stdin` is type of `*os.File`, while the `ReadPassword` accepts `int` type as argument. I tried to use `syscall.Stdin`, but it has compilation problem, would have error: 

```
cannot use syscall.Stdin (type syscall.Handle) as type int in argument to terminal.ReadPassword
```

Comment (gz-c) [reply src/api/cli/cli.go]: Nevermind, its safe to use 0, syscall.Stdin is always 0 anyway

Comment (iketheadore) [reply src/api/cli/cli.go]: Yes, syscall.Stdin is 0, but not for windows, as the `build` flag in `syscall_unix.go` is
`// +build darwin dragonfly freebsd linux netbsd openbsd solaris`.

Comment (gz-c) [reply src/api/cli/cli.go]: What is it in windows? Then do we need to do `int(syscall.Stdin)`?

Comment (gz-c) [reply src/api/cli/cli.go]: I see, `syscall_windows.go:	Stdin  = getStdHandle(STD_INPUT_HANDLE)`

then we probably need to do `int(os.Stdin.Fd())` or `int(syscall.Stdin)`

File: src/api/cli/add_private_key.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gz-c) [reply src/api/cli/add_private_key.go]: should still check in case that changes

File: src/api/cli/cli.go (hunk 1 above)
Comment (iketheadore) [reply src/api/cli/cli.go]: I will test `int(syscall.Stdin)` on windows to see if it is able to work

Comment (iketheadore) [reply src/api/cli/cli.go]: OK,` int(syscall.Stdin)` works, I will use it.

**Pull request merged**